expression/substring.go

```diff
@@ -112,3 +118,86 @@ func (f *FunctionSubstring) Eval(ctx context.Context, args map[interface{}]inter
 func (f *FunctionSubstring) Accept(v Visitor) (Expression, error) {
 	return v.VisitFunctionSubstring(f)
 }
+
+// FunctionSubstringIndex returns the substring as specified.
+// See: https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_substring-index
+type FunctionSubstringIndex struct {
+	StrExpr Expression
+	Delim   Expression
+	Count   Expression
+}
+
+// Clone implements the Expression Clone interface.
+func (f *FunctionSubstringIndex) Clone() Expression {
+	nf := &FunctionSubstringIndex{
+		StrExpr: f.StrExpr.Clone(),
+		Delim:   f.Delim.Clone(),
+		Count:   f.Count.Clone(),
+	}
+	return nf
+}
+
+// IsStatic implements the Expression IsStatic interface.
+func (f *FunctionSubstringIndex) IsStatic() bool {
+	return f.StrExpr.IsStatic() && f.Delim.IsStatic() && f.Count.IsStatic()
+}
+
+// String implements the Expression String interface.
+func (f *FunctionSubstringIndex) String() string {
+	return fmt.Sprintf("SUBSTRING_INDEX(%s, %s, %s)", f.StrExpr.String(), f.Delim.String(), f.Count.String())
+}
+
+// Eval implements the Expression Eval interface.
+func (f *FunctionSubstringIndex) Eval(ctx context.Context, args map[interface{}]interface{}) (interface{}, error) {
+	fs, err := f.StrExpr.Eval(ctx, args)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	str, ok := fs.(string)
+	if !ok {
+		return nil, errors.Errorf("Substring_Index invalid args, need string but get %T", fs)
+	}
+
+	t, err := f.Delim.Eval(ctx, args)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	delim, ok := t.(string)
+	if !ok {
+		return nil, errors.Errorf("Substring_Index invalid delim, need string but get %T", t)
+	}
+
+	t, err = f.Count.Eval(ctx, args)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	c, ok := t.(int64)
+	if !ok {
+		return nil, errors.Errorf("Substring invalid pos args, need int but get %T", t)
+	}
+	count := int(c)
+	strs := strings.Split(str, delim)
+	var (
+		start = 0
+		end   = len(strs)
+	)
+	if count > 0 {
+		// If count is positive, everything to the left of the final delimiter (counting from the left) is returned.
+		if count < end {
+			end = count
+		}
+	} else {
+		// If count is negative, everything to the right of the final delimiter (counting from the right) is returned.
+		count = -count
+		if count < end {
+			start = end - count
+		}
+	}
+	substrs := strs[start:end]
```

Comment:
start init to 0 and will not be greater than len(strs) or small than 0
end init to len(strs) and will not be greater than len(strs) or smaller than 0

And start will be aways less or equal than end.